### PR TITLE
Factor DrawerController out of Drawer

### PR DESCRIPTION
**Pull request by abarth**
This patch paves the path for refactoring the Drawer into the Scaffold to
support edge swipes and peeks.

### CODE REVIEW

File: sky/packages/sky/lib/src/material/drawer.dart [hunk 1]
```diff
@@ -24,11 +23,7 @@ import 'theme.dart';
 
 const double _kWidth = 304.0;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/material/drawer.dart]: add unit comments to these constants

File: sky/packages/sky/lib/src/material/drawer.dart [hunk 2]
```diff
@@ -141,9 +175,49 @@ class _DrawerRoute extends TransitionRoute {
...
+
+  Widget build(BuildContext context) {
+    _performance.updateVariable(_color);
+    Matrix4 transform = new Matrix4.identity()
+      ..translate((-1.0 + _performance.progress) * (_width ?? 0.0));
+    Widget result = new GestureDetector(
+      onHorizontalDragUpdate: _move,
+      onHorizontalDragEnd: _settle,
+      child: new Stack(<Widget>[
+        new GestureDetector(
```
Comment (Hixie) [inline sky/packages/sky/lib/src/material/drawer.dart]: Why not a ModalBarrier?

File: sky/packages/sky/lib/src/material/drawer.dart [hunk 3]
```diff
@@ -38,101 +33,140 @@ class _Drawer extends StatelessComponent {
...
+  }
+
+  void _performanceChanged() {
+    setState(() {
+      // The performance's state is our build state, and it changed already.
+    });
+  }
+
+  void _performanceStatusChanged(PerformanceStatus status) {
+    print('_performanceStatusChanged $status');
```
Comment (Hixie) [inline sky/packages/sky/lib/src/material/drawer.dart]: extraneous print

**Comment by Hixie** [comment]
LGTM, but travis is mad

File: sky/packages/sky/lib/src/material/drawer.dart (hunk 2 above)
Comment (abarth) [inline sky/packages/sky/lib/src/material/drawer.dart]: ModalBarrier pops the route on down rather than tap.  Maybe it should be configurable?  Maybe we should use tap everywhere?

Comment (Hixie) [inline sky/packages/sky/lib/src/material/drawer.dart]: Yeah let's use tap everywhere. That way you can avoid cancelling e.g. a dialog if you move out of the slop.

Comment (abarth) [inline sky/packages/sky/lib/src/material/drawer.dart]: Ok.  Will do in a separate patch.  Currently switching this patch to use Align instead of Transform, which gets rid of the Offstage nonsense.

Comment (Hixie) [inline sky/packages/sky/lib/src/material/drawer.dart]: Sounds good.

File: sky/packages/sky/lib/src/rendering/shifted_box.dart [hunk 4]
```diff
@@ -197,11 +197,9 @@ class RenderPositionedBox extends RenderShiftedBox {
...
     if (child != null) {
       child.layout(constraints.loosen(), parentUsesSize: true);
-      final Size desiredSize = new Size(child.size.width * (_widthFactor ?? 1.0),
-                                        child.size.height * (_heightFactor ?? 1.0));
-      size = constraints.constrain(new Size(shrinkWrapWidth ? desiredSize.width : double.INFINITY,
-                                            shrinkWrapHeight ? desiredSize.height : double.INFINITY));
-      final Offset delta = size - desiredSize;
+      size = constraints.constrain(new Size(shrinkWrapWidth ? child.size.width * _widthFactor : double.INFINITY,
+                                            shrinkWrapHeight ? child.size.height * _heightFactor : double.INFINITY));
+      final Offset delta = size - child.size;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/shifted_box.dart]: I tweaked this logic a bit to make it possible to align things to the right edge.

Comment (HansMuller) [inline sky/packages/sky/lib/src/rendering/shifted_box.dart]: That's the way it should have been. Thanks!
